File: DynamicProgramming/5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 *
 * https://leetcode.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (27.87%)
 * Likes:    4495
 * Dislikes: 406
 * Total Accepted:    677K
 * Total Submissions: 2.4M
 * Testcase Example:  '"babad"'
 *
 * Given a string s, find the longest palindromic substring in s. You may
 * assume that the maximum length of s is 1000.
 * 
 * Example 1:
 * 
 * 
 * Input: "babad"
 * Output: "bab"
 * Note: "aba" is also a valid answer.
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "cbbd"
 * Output: "bb"
 * 
 * 题目大意，找数组中最长回文子串
 * 解题思路1：
 * 所有可能的子串个数为n^2个，想不出来o(n)的解，所有初步思路是处理每一个可能
 * 的子串。对于单个子串如果用普通的判断是否回文的方法可能要到n^3级别，因此思路
 * 就是利用以前的信息（动态规划），对于每个串用o(1)的时间判断是否回文。
 * 1、构建一个上三角的矩阵，初值全为0，表示第j行到第i列的不回文（j<=i）
 * 2、对角线全设置为1
 * 3、处理每一列
 *  如果s[i]!=s[j],不可能回文，设0；
 *  如果s[i]==s[j]，要看中间的会不会回文，即sj+1-si-1是不是回文的，
 *      a.如果是j+1==i，表示当前串只有两个元素，sjsi(j+1=i)，
 *      二者有相等，肯定回文
 *      b.否则，看即sj+1-si-1是否回文reocrd[j+1][i-1]==1，肯定回文，
 *      不回文否则
 * 解法2：最长公共子串的应用
 * 找s和reverse的最长公共子串，但是由于最长公共子串本身可能是不回文的，
 * 所以找到之后要判断是否回文，不回文的话要找次长的
 * 解法3：暴力
 * 时间O(N^3)，空间O(1)
 * 解法4：动态规划的空间提升
 * 由解法1而来，其空间为平方级别，但是每次只用了reocrd[j+1][i-1]的结果，
 * 因此可以只保存一列，而不是所有的矩阵，空间O（n）
 * 解法5：动态规划的空间提升2
 * 作为回文肯定有一个center，这个center可能是单个字符（奇数串），也可能是
 * 两个字符（偶数串），而这样的center一共就有n+(n-1)=2n-1个。
 * 可以对这2n-1个中心，依次处理，找到其最长的回文串并记录.
 * 时间o(n^2)空间O(1)
 * 解法5：manacher's算法
 * 时间o(n)
 * 在解法5的基础上。
 * 1、先在每个元素之间插入#，这样就都可以作为奇数长的回文串处理。
 * 2、用R记录当前最长的回文串所能达到的最远下标；C记录当前最长回文串的中心元素
 * 所在下标；R=-1,C=-1;数组z记录回文半径
 * 3、从左到右进行更新：
 *   if i>R，从i开始向左右扩展，同时更新R和C；
 *   else   
 *      计算当前最长回文串的左边界CL，i关于C对称的元素i'对于的左边界；
 *      if cl<pl，则根据i和i'互为对称可知z[i]=z[i']
 *      if cl>pl，则为R-i+1；
 *      if cl==pl，则要从R往两边扩展
 */

// @lc code=start
class Solution {
public:
string longestPalindrome(string s) {
    string ss="#";
    for(int i=0;i<s.size();i++){
        ss+=s[i];
        ss+="#";
    }
    vector<int>rd(ss.size(),0);
    int R=-1,C=-1;
    int maxr=0;
    string rets="";
    for(int i=0;i<ss.size();i++){
        if(R<=i){
            rd[i]=1;
        }else{
            rd[i]=min(rd[2*C-i],R-i+1);
        }
        while(i+rd[i]<ss.size()&&i-rd[i]>-1){
            if(ss[i-rd[i]]==ss[i+rd[i]]){
                rd[i]++;
            }else{
                break;
            }
        }
        if(i+rd[i]>R){
            R=i+rd[i]-1;
            C=i;
        }
        
        if(maxr<rd[i]){
            rets=ss.substr(i-rd[i]+1,2*rd[i]-1);
            maxr=rd[i];
        }
    }
    string ret="";
    for(int i=0;i<rets.size();i++){
        if(rets[i]!='#'){
            ret+=rets[i];
        }
    }
    return ret;
}
string longestPalindrome4(string s) {
    if(s.empty())return s;
    if(s.size()==1)return s;
    int n=s.size();
    int start=0,end=0;
    for(int i=0;i<n;i++){
        int len1=expandAroundCenter(s,i,i);
        int len2=expandAroundCenter(s,i,i+1);
        int len=max(len1,len2);
        if(len>end-start){
            start=i-(len-1)/2;
            end=i+len/2;
        }
    }
    return s.substr(start,end-start+1);
}
int expandAroundCenter(string s,int left,int right){
    int L=left,R=right;
    while(L>=0&&R<s.size()&&s[L]==s[R]){
        L--;
        R++;
    }
    return R-L-1;
}
    string longestPalindrome1(string s) {
        if(s.empty())return s;
        if(s.size()==1)return s;
        int n=s.size();
        string ret="";
        vector<int>record(n,0);
        for(int i=0;i<n;i++){
            for(int j=0;j<=i;j++){
                if(i==j){
                    record[j]=1;
                }else{
                    if(s[i]==s[j]){
                        //相等才可能回文
                        if(j+1<i&&record[j+1]==1){
                            record[j]=1;
                        }else if(j+1>=i){
                            record[j]=1;
                        }else{
                            record[j]=0;
                        }
                    }else{
                        record[j]=0;
                    }
                    
                }
                if(record[j]==1){
                    string tmp=s.substr(j,i-j+1);
                    if(tmp.size()>ret.size()){
                        ret.assign(tmp);
                    }
                }
            }
            
        }
        return ret;
    }
};
// @lc code=end

File: String/5.longest-palindromic-substring.java
import sun.security.util.Length;

/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 *
 * https://leetcode.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (27.87%)
 * Likes:    4495
 * Dislikes: 406
 * Total Accepted:    677K
 * Total Submissions: 2.4M
 * Testcase Example:  '"babad"'
 *
 * Given a string s, find the longest palindromic substring in s. You may
 * assume that the maximum length of s is 1000.
 * 
 * Example 1:
 * 
 * 
 * Input: "babad"
 * Output: "bab"
 * Note: "aba" is also a valid answer.
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "cbbd"
 * Output: "bb"
 * 
 * 注意最长公共子串（要求更高，转化为最长公共后缀的长度问题）和最长公共子序列的区别
 * 不仅是最长公共子串，还是最长公共回文子串，所以会失效
 * 要考虑：We could see that the longest common substring method fails 
 * when there exists a reversed copy of a non-palindromic substring 
 * in some other part of SS.
 * 
 * 解题思路：
 * 1、求s和s.reverse()的最长的回文的公共子串
 * T:O(n^2)和S:O(N^2)
 * 2、Brute Force
 * T:O(n^3)和S:O(1)
 * 3、中心扩展法
 * 对于每个元素i，像左右两边扩展，计算得到以i为中心的最长回文串，并更新最大
 * 注意对每个元素有两种扩展方法，一种是回文串为偶数长度，一种是为奇数长度
 * T:O(n^2)和S:O(1)
 * 4、动态规划
 * 一般的动态规划是二维数组从左上角开始往左下角填；
 * 本题的动态规划是上三角矩阵，逐条对角线，从左上角往右下角计算
 * dp[i,j]表示sub(i,j+1)是否是回文串，那么
 * dp[i,j]==true，当dp[i+1,j-1]&&s[i]==s[j]
 * 否则dp[i,j]==false,
 * 初始情况是dp[i][i]=true,dp[i][i+1]=(s[i]==s[i+1])
 * T:O(n^2)和S:O(N^2)
 */

// @lc code=start
class Solution {
    /**
     * 动态规划法
     * 
     * @param s
     * @return
     */
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        if (s.length() == 1) {
            return s;
        }
        int n = s.length();
        String ans = "";
        boolean[][] dp = new boolean[n][n];
        for (int i = 0; i < n; i++) {
            dp[i][i] = true;
            if (i + 1 < n) {
                dp[i][i + 1] = (s.charAt(i) == s.charAt(i + 1));
                if (dp[i][i + 1] && s.substring(i, i + 2).length() > ans.length()) {
                    ans = s.substring(i, i + 2);
                }
            }
            if (dp[i][i] && s.substring(i, i + 1).length() > ans.length()) {
                ans = s.substring(i, i + 1);
            }

        }
        for (int k = 2; k < n; k++) {
            for (int j = k, i = 0; i < n && j < n; i++, j++) {
                dp[i][j] = (dp[i + 1][j - 1] && s.charAt(i) == s.charAt(j));
                if (dp[i][j] && s.substring(i, j + 1).length() > ans.length()) {
                    ans = s.substring(i, j + 1);
                }
            }
        }
        return ans;
    }

    /**
     * 中心扩展法
     * 
     * @param s
     * @return
     */
    public String longestPalindrome3(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        if (s.length() == 1) {
            return s;
        }
        int n = s.length();
        String ans = "";
        for (int i = 0; i < n; i++) {
            // 循环不变量i表示中心点范围[0,n-1]
            // 奇数，以s[i]为中心
            int left = i - 1, right = i + 1;// left=[-1,i-1],right=[i+1,n-1]
            while (left >= 0 && right < n) {
                if (s.charAt(left) != s.charAt(right)) {
                    break;
                }
                left--;
                right++;
            }
            // 一定有结果至少是单个字符
            if (s.substring(left + 1, right).length() > ans.length()) {
                ans = s.substring(left + 1, right);
            }
            // 偶数，以s[i]s[i+1]为中心
            left = i;// left=[0,i]
            right = i + 1;// right=[i+1,n-1]
            while (left >= 0 && right < n) {
                if (s.charAt(left) != s.charAt(right)) {
                    break;
                }
                left--;
                right++;
            }
            // 不一定有结果的
            if (left != i && s.substring(left + 1, right).length() > ans.length()) {
                // left动了，表示找到了至少一个回文串
                ans = s.substring(left + 1, right);
            }
        }
        return ans;
    }

    /**
     * Brute Force
     * 
     * @param s
     * @return
     */
    public String longestPalindrome2(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        if (s.length() == 1) {
            return s;
        }
        int n = s.length();
        String ans = "";
        // i,j循环不变量，表示回文串s[i,j)，所以i范围是[0,n-1],j范围是[i+1,n]
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j <= n; j++) {
                String sub = s.substring(i, j);
                if (isPalindrome(sub) && sub.length() > ans.length()) {
                    ans = sub;
                }
            }
        }
        return ans;
    }

    public boolean isPalindrome(String s) {
        if (s == "" || s.length() == 1) {
            return true;
        }
        int i = 0, j = s.length() - 1;// 循环不变量，表示比较的元素
        while (i < j) {// i==j不用判断了
            if (s.charAt(i) != s.charAt(j)) {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }

    /**
     * 写错的使用公共子串方法计算最长回文子串： 应该改为对于求得的最长回文子串还要判断它是否回文，找到回文的最长的 那个才是答案。
     * 
     * 对于求最长公共子串的方法要注意，状态转移方程 当s[i]!=s[j]的时候，dp[j][j]=0，而不是max(,)
     * 
     * @param s
     * @return
     */
    public String longestPalindrome1(String s) {
        if (s == null || s.length() == 0) {
            return "";
        }
        if (s.length() == 1) {
            return s;
        }
        int n = s.length();
        int[][] dp = new int[n + 1][n + 1];
        ;
        String b = new StringBuilder(s).reverse().toString();
        int maxLen = 0, indi = -1, indj = -1;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (s.charAt(i - 1) == b.charAt(j - 1)) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {
                    dp[i][j] = 0;// 这个是最长公共子串的写法
                    // dp[i][j]=Math.max(dp[i][j-1], dp[i-1][j]);这个是最长公共子序列的解法
                }
                if (dp[i][j] > maxLen) {
                    maxLen = dp[i][j];
                    indi = i;
                    indj = j;
                }
            }
        }
        // System.out.print(dp[n][n]);
        StringBuilder ans = new StringBuilder();
        if (maxLen == 0) {
            return "";
        }
        while (ans.length() < maxLen && indi - 1 < n && indj - 1 < n) {

            ans.append(s.charAt(indi - 1));
            indi--;
            indj--;

        }
        return ans.toString();
    }
}
// @lc code=end
